### PR TITLE
Flow properties first pass

### PR DESCRIPTION
**Pull request by MingweiSamuel**
#271

### CODE REVIEW

File: hydroflow_lang/src/graph/propegate_flow_props.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                    // TODO(mingwei): don't exit early here?
+                    let flow_props_out =
+                        (flow_prop_fn)(flow_prop_args, diagnostics).map_err(|()| node_id)?;
+                    assert!(
+                        1 == flow_props_out.len()
+                            || graph.node_degree_out(node_id) == flow_props_out.len()
+                    );
+
+                    let out_edges = graph.node_successor_edges(node_id).collect::<Vec<_>>();
+                    // TODO(mingwei): SORT EDGES DETERMINISTICALLY BY NAME (SOMEHOW?).
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/propegate_flow_props.rs:48]: do this

File: hydroflow_lang/src/graph/propegate_flow_props.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                }
+            }
+            Node::Handoff { .. } => {
+                assert_eq!(1, graph.node_degree_in(node_id));
+                assert_eq!(1, graph.node_degree_out(node_id));
+                let in_edge = graph.node_predecessor_edges(node_id).next().unwrap();
+                let out_edge = graph.node_successor_edges(node_id).next().unwrap();
+                if let Some(flow_props) = graph.edge_flow_props(in_edge) {
+                    graph.set_edge_flow_props(out_edge, flow_props);
+                }
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/propegate_flow_props.rs:66]: handoffs just repeat their input flow props

File: hydroflow_macro/src/lib.rs [hunk 3]
```diff
@@ -154,6 +154,20 @@ fn hydroflow_wrap(item: proc_macro::TokenStream, attribute: Attribute) -> proc_m
...
+    item
+}
+
+/// Christen the given closure as a monotonic function.
+#[proc_macro]
+pub fn monotonic_fn(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    // TODO(mingwei): some sort of code analysis?
+    item
+}
+
```
Comment (MingweiSamuel) [inline hydroflow_macro/src/lib.rs:170]: Should these be `#[doc(hidden)]`?

File: hydroflow_lang/src/graph/ops/mod.rs [hunk 4]
```diff
@@ -390,6 +469,31 @@ impl WriteContextArgs<'_> {
...
+            .and_then(|flow_props| flow_props.lattice_flow_type)
+        {
+            None => quote_spanned!(span=> #func_arg),
+            Some(LatticeFlowType::Cumul) => quote_spanned! {span=>
+                #root::monotonic_fn!(#func_arg)
+            },
+            Some(LatticeFlowType::Delta) => quote_spanned! {span=>
+                #root::morphism!(#func_arg)
+            },
+        }
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/ops/mod.rs:495]: this doesn't do actual type checking yet

**Review by zzlk — APPROVED** [review]

File: hydroflow_macro/src/lib.rs (hunk 3 above)
Comment (MingweiSamuel) [reply hydroflow_macro/src/lib.rs:170]: maybe, but no for now

File: hydroflow_lang/src/graph/propegate_flow_props.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+use crate::diagnostic::Diagnostic;
+use crate::graph::graph_algorithms;
+use crate::graph::ops::FlowPropArgs;
+
+/// Traverses the graph, propegating the flow properties from sources to sinks.
+pub fn propegate_flow_props(
+    graph: &mut HydroflowGraph,
+    diagnostics: &mut Vec<Diagnostic>,
+) -> Result<(), GraphNodeId> {
+    let node_order = graph_algorithms::topo_sort_scc(
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/propegate_flow_props.rs:15]: will merge this now and fix as #895 later

File: hydroflow_lang/src/graph/propegate_flow_props.rs (hunk 1 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/propegate_flow_props.rs:55]: Not needed, the edges are in the same order as the `output_ports` in `op_inst`

**Pull request merged**